Sources/Core/GraphRequest.swift

```diff
@@ -0,0 +1,160 @@
+     - Parameter tokenString: an optional access token to use, must provide a token for paths that require a token
+     - Parameter version: the optional Graph API version (e.g., "v2.0"). nil defaults to `Settings.graphAPIVersion`.
+     - Parameter httpMethod: the HTTP method. Empty String defaults to `HTTPMethod.get`
+  */
+  init(
+    graphPath: GraphPath,
+    parameters: [String: Any] = [:],
+    accessToken: AccessToken? = AccessTokenWallet.shared.currentAccessToken,
+    version: String = Settings.graphAPIVersion,
+    httpMethod: GraphRequest.HTTPMethod = .get,
+    flags: GraphRequest.Flags = .none,
+    enableGraphRecovery: Bool = Settings.isGraphErrorRecoveryEnabled
+    ) {
+    self.graphPath = graphPath
+    self.parameters = parameters
+    self.accessToken = accessToken
+    self.version = version
+    self.httpMethod = httpMethod
+
+    var flags = flags
+    if !enableGraphRecovery {
+      flags = .disableErrorRecovery
+    }
+    self.flags = flags
+  }
+
+  var isGraphRecoveryDisabled: Bool {
+    return flags.contains(.disableErrorRecovery)
+  }
+
+  /**
+   Enable or disable the automatic error recovery mechanism.
+
+   - Parameter enabled: whether to enable the automatic error recovery mechanism
+
+   By default, non-batched GraphRequest instances will automatically try to recover
+   from errors by constructing a `GraphErrorRecoveryProcessor` instance that
+   re-issues the request on successful recoveries. The re-issued request will call the same
+   handler as the receiver but may occur with a different `GraphRequestConnection` instance.
+
+   This will override `Settings.setGraphErrorRecoveryDisabled`
+  */
+  mutating func setGraphErrorRecoverability(enabled: Bool) {
+    if enabled {
+      flags.remove(.disableErrorRecovery)
+    } else {
+      flags.insert(.disableErrorRecovery)
+    }
+  }
+
+  /**
+   Start the graph request on a `GraphRequestConnection`
+
+   - Parameter with: a connection to begin the request on. Generally a best practice to omit this parameter
+   and allow the request to provide a new instance of a connection
+   - Parameter completion: A handler for when the `GraphRequestConnection` completes the `GraphRequest`
+
+   - Returns: An object that conforms to `GraphRequestConnecting` and is executing the `GraphRequest`
+  */
+  func start(with connection: GraphRequestConnecting = GraphRequestConnection(),
+             completion handler: @escaping GraphRequestBlock) -> GraphRequestConnecting {
+    connection.add(request: self, completion: handler)
+    connection.start()
+    return connection
+  }
+
+  /// Returns true if any of the parameters are of type `UIImage`, `Data` or `GraphRequestDataAttachment`
+  var hasAttachments: Bool {
+    for (_, item) in parameters {
+      if GraphRequest.isAttachment(item) {
+        return true
+      }
+    }
+    return false
+  }
+
+  private static func isAttachment(_ item: Any) -> Bool {
+    return item is UIImage ||
+      item is Data ||
```

Comment:
Better blame?

```suggestion
      || item is Data
```

Reply:
I think it's about the same. I don't find one more readable particularly than the other so happy to change it. :)